src/vs/workbench/services/textMate/browser/worker/textMateWorkerModel.ts

```diff
@@ -0,0 +1,161 @@
+		this._isDisposed = true;
+		super.dispose();
+	}
+
+	public onLanguageId(languageId: string, encodedLanguageId: LanguageId): void {
+		this._languageId = languageId;
+		this._encodedLanguageId = encodedLanguageId;
+		this._resetTokenization();
+	}
+
+	private readonly tokenizeDebouncer = new RunOnceScheduler(() => this._tokenize(), 10);
+
+	override onEvents(e: IModelChangedEvent): void {
+		super.onEvents(e);
+		if (this._tokenizationStateStore) {
+			for (let i = e.changes.length - 1; i >= 0; i--) {
+				const change = e.changes[i];
+				const [eolCount] = countEOL(change.text);
+				this._tokenizationStateStore.applyEdits(change.range, eolCount);
+			}
+		}
+		this.tokenizeDebouncer.schedule();
+	}
+
+	public retokenize(startLineNumber: number, endLineNumberExclusive: number) {
+		if (this._tokenizationStateStore) {
+			for (let lineNumber = startLineNumber; lineNumber < endLineNumberExclusive; lineNumber++) {
+				this._tokenizationStateStore.markMustBeTokenized(lineNumber - 1);
+			}
+			this.tokenizeDebouncer.schedule();
+		}
+	}
+
+	private _resetTokenization(): void {
+		this._tokenizationStateStore = null;
+
+		const languageId = this._languageId;
+		const encodedLanguageId = this._encodedLanguageId;
+		this._worker.getOrCreateGrammar(languageId, encodedLanguageId).then((r) => {
+			if (this._isDisposed || languageId !== this._languageId || encodedLanguageId !== this._encodedLanguageId || !r) {
+				return;
+			}
+
+			if (r.grammar) {
+				const tokenizationSupport = new TextMateTokenizationSupport(r.grammar, r.initialState, false);
+				this._tokenizationStateStore = new TokenizationStateStore(tokenizationSupport, tokenizationSupport.getInitialState());
+			} else {
+				this._tokenizationStateStore = null;
+			}
+			this._tokenize();
+		});
+	}
+
+	private _tokenize(): void {
+		if (!this._tokenizationStateStore) {
+			return;
+		}
+
+		const startTime = new Date().getTime();
+
+		while (true) {
+			const builder = new ContiguousMultilineTokensBuilder();
+			const lineCount = this._lines.length;
+
+			let tokenizedLines = 0;
+
+			const stateDeltaBuilder = new StateDeltaBuilder();
+
+			// Validate all states up to and including endLineIndex
+			while (this._tokenizationStateStore.invalidLineStartIndex < lineCount) {
+				const lineIndex = this._tokenizationStateStore.invalidLineStartIndex;
+
+				tokenizedLines++;
+				// TODO don't spam the renderer
+				if (tokenizedLines > 200) {
+					break;
+				}
+
+				const text = this._lines[lineIndex];
+
+				const lineStartState = this._tokenizationStateStore.getBeginState(lineIndex) as StateStack;
+				const tokenizeResult = this._tokenizationStateStore.tokenizationSupport.tokenizeEncoded(text, true, lineStartState);
+				if (this._tokenizationStateStore.setEndState(lineCount, lineIndex, tokenizeResult.endState)) {
+					const delta = diffStateStacksRefEq(lineStartState, tokenizeResult.endState as StateStack);
+					stateDeltaBuilder.setState(lineIndex + 1, delta);
+				}
+
```

Comment:
I know the old code didn't have it, but we should skip tokenizing long lines. It would be great if we could wire in `editor.maxTokenizationLineLength` to the textmate web worker.

Reply:
Let's do this in a separate PR.